### ISSUE TITLE
"and not" matching expression not possible from secondary master in multi-master configuration

### ISSUE THREAD

**Issue by deuscapturus** [saltstack/salt#16435]
salt-2014.1.10-4.el6.noarch

From primary salt master  (first in list)
```
[root@ip-10-1-0-116 ec2-user]# salt -C '* and not G@deployment:production' test.ping
ip-10-1-0-25.ec2.internal:
    True
ip-10-1-0-77.ec2.internal:
    True
ip-172-31-61-203.us-west-2.compute.internal:
    True
ip-172-31-54-209.us-west-2.compute.internal:
    True
ip-172-31-54-104.us-west-2.compute.internal:
    True
ip-172-31-23-201.us-west-2.compute.internal:
    True
```

From secondary salt master (second in list)
```
[root@ip-172-31-11-86 minions]# salt -C '* and not G@deployment:production' test.ping
No minions matched the target. No command was sent, no jid was assigned.
```

I've noticed that no minion cache exists in the secondary salt master.  And I have not found a way to generate the minion cache on secondary master.

From secondary salt master after running `salt '*' saltutil.sync_all`
```
[root@ip-172-31-11-86 minions]# ls -l /var/cache/salt/master/minions
total 0
[root@ip-172-31-11-86 minions]# salt-run cache.grains
[WARNING ] DEPRECATION WARNING: The 'minion' arg will be removed from cache.py runner. Specify minion with 'tgt' arg!
ip-10-1-0-114.ec2.internal:
    ----------
ip-10-1-0-201.ec2.internal:
    ----------
ip-10-1-0-25.ec2.internal:
    ----------
ip-10-1-0-250.ec2.internal:
    ----------
ip-10-1-10-115.ec2.internal:
    ----------
ip-10-1-10-171.ec2.internal:
    ----------
ip-10-1-10-197.ec2.internal:
    ----------
ip-10-1-12-233.ec2.internal:
    ----------
ip-10-1-13-38.ec2.internal:
    ----------
ip-172-31-23-201.us-west-2.compute.internal:
    ----------
ip-172-31-54-104.us-west-2.compute.internal:
    ----------
ip-172-31-54-209.us-west-2.compute.internal:
    ----------
ip-172-31-61-203.us-west-2.compute.internal:
    ----------
ip-172-31-63-39:
    ----------
```

**Comment by cachedout** [comment]
I have tested this against the HEAD of 2014.7 and it works correctly.

```
[root@21744_master1 ~]# salt -C '* and not G@foo:bar' test.ping
21744_minion1:
    True
```

and

```
[root@21744_master2 ~]# salt -C '* and not G@foo:bar' test.ping
21744_minion1:
    True
```

I believe this issue has been resolved. I am going to tag it accordingly..

**Comment by rallytime** [comment]
@deuscapturus Can you give the HEAD of 2014.7 a try and let us know if this is fixed for you or not?

**Comment by rallytime** [comment]
Since we haven't heard back for a bit and this looks to be fixed, I'm going to close this issue. This fix should be in 2014.7.5. Please leave a comment if this comes up again and we'll open this issue again and readdress!
